auth0/src/main/java/com/auth0/android/authentication/storage/CredentialsManager.java

```diff
@@ -0,0 +1,108 @@
+     *
+     * @param authenticationClient the Auth0 Authentication client to refresh credentials with.
+     * @param storage              the storage to use for the credentials.
+     */
+    public CredentialsManager(@NonNull AuthenticationAPIClient authenticationClient, @NonNull Storage storage) {
+        this.authClient = authenticationClient;
+        this.storage = storage;
+    }
+
+    /**
+     * Stores the given credentials in the storage. Must have an access_token or id_token and a expires_in value.
+     *
+     * @param credentials the credentials to save in the storage.
+     */
+    public void setCredentials(@NonNull Credentials credentials) {
+        if ((isEmpty(credentials.getAccessToken()) && isEmpty(credentials.getIdToken())) || credentials.getExpiresIn() == null) {
+            throw new CredentialsManagerException("Credentials must have a valid expires_in value and a valid access_token or id_token value.");
+        }
+        storage.store(KEY_ACCESS_TOKEN, credentials.getAccessToken());
+        storage.store(KEY_REFRESH_TOKEN, credentials.getRefreshToken());
+        storage.store(KEY_ID_TOKEN, credentials.getIdToken());
+        storage.store(KEY_TOKEN_TYPE, credentials.getType());
+
+        Long expiresIn = credentials.getExpiresIn();
+        storage.store(KEY_EXPIRES_IN, Long.toString(expiresIn));
+        long expirationTime = getCurrentTimeInMillis() + (expiresIn * 1000);
+        storage.store(KEY_EXPIRATION_TIME, Long.toString(expirationTime));
+    }
+
+    /**
+     * Retrieves the credentials from the storage and refresh them if they have already expired.
+     * It will fail with {@link CredentialsManagerException} if the saved access_token or id_token is null,
+     * or if the tokens have already expired and the refresh_token is null.
+     *
+     * @param callback the callback that will receive a valid {@link Credentials} or the {@link CredentialsManagerException}.
+     */
+    public void getCredentials(@NonNull final BaseCallback<Credentials, CredentialsManagerException> callback) {
+        String accessToken = storage.retrieve(KEY_ACCESS_TOKEN);
+        String refreshToken = storage.retrieve(KEY_REFRESH_TOKEN);
+        String idToken = storage.retrieve(KEY_ID_TOKEN);
+        String tokenType = storage.retrieve(KEY_TOKEN_TYPE);
+        String expiresInValue = storage.retrieve(KEY_EXPIRES_IN);
+        String expirationTimeValue = storage.retrieve(KEY_EXPIRATION_TIME);
+
+        if (isEmpty(accessToken) && isEmpty(idToken) || isEmpty(expiresInValue) || isEmpty(expirationTimeValue)) {
+            callback.onFailure(new CredentialsManagerException("No Credentials were previously set."));
+            return;
+        }
+        long expirationTime = Long.parseLong(expirationTimeValue);
+        if (expirationTime > getCurrentTimeInMillis()) {
+            Long expiresIn = Long.parseLong(expiresInValue);
```

Comment:
the comment should be here ... sorry I commented with the link on the email and it comments on the commit not in the general PR